src/libraries/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs

```diff
@@ -56,32 +58,120 @@ public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mo
-                ewh.SafeWaitHandle = handle;
-                old.Dispose();
+                if (handle.IsInvalid)
+                {
+                    handle.SetHandleAsInvalid();
+
+                    if (!string.IsNullOrEmpty(name) && errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
+                    {
+                        throw new WaitHandleCannotBeOpenedException(SR.Format(SR.WaitHandleCannotBeOpenedException_InvalidHandle, name));
+                    }
+
+                    throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+                }
+
+                createdNew = (errorCode != Interop.Errors.ERROR_ALREADY_EXISTS);
 
-                return ewh;
+                return CreateAndReplaceHandle(handle);
             }
         }
 
-        private static void ValidateHandle(SafeWaitHandle handle, string? name, out bool createdNew)
+        /// <summary>
+        /// Opens a specified named event wait handle, if it already exists, applying the desired access rights.
+        /// </summary>
+        /// <param name="name">The name of the event wait handle to be opened. If it's prefixed by "Global", it refers to a machine-wide event wait handle. If it's prefixed by "Local", or doesn't have a prefix, it refers to a session-wide event wait handle. Both prefix and name are case-sensitive.</param>
+        /// <param name="rights">The desired access rights to apply to the returned event wait handle.</param>
+        /// <returns>An existing named event wait handle.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="name"/> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentException"><paramref name="name"/> is an empty string.</exception>
+        /// <exception cref="WaitHandleCannotBeOpenedException">The named event wait handle does not exist or is invalid.</exception>
+        /// <exception cref="IOException">The path was not found.
+        /// -or-
+        /// A Win32 error occurred.</exception>
+        /// <exception cref="UnauthorizedAccessException">The named event wait handle exists, but the user does not have the security access required to use it.</exception>
+        public static EventWaitHandle OpenExisting(string name, EventWaitHandleRights rights)
         {
-            int errorCode = Marshal.GetLastWin32Error();
+            switch (OpenExistingWorker(name, rights, out EventWaitHandle? result))
+            {
+                case OpenExistingResult.NameNotFound:
+                    throw new WaitHandleCannotBeOpenedException();
+
+                case OpenExistingResult.NameInvalid:
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
 
-            if (handle.IsInvalid)
+                case OpenExistingResult.PathNotFound:
+                    throw new IOException(SR.Format(SR.IO_PathNotFound_Path, name));
```

Comment:
Good catch, I'll fix it.
Speaking of this particular exception - I haven't been able to find a case that can throw the exception inside this condition, so I haven't added a unit test for it yet. Do you have any suggestions on what name to pass so that we throw this, @kouvel ?

Reply:
Note:
Only Mutex and EventWaitHandle throw `DirectoryNotFoundException` in the other overloads:
https://source.dot.net/#System.Private.CoreLib/Mutex.cs,50
https://source.dot.net/#System.Private.CoreLib/EventWaitHandle.cs,41

Semaphore throws IOException:
https://source.dot.net/#System.Private.CoreLib/Semaphore.cs,46

Reply:
You're right, it looks like `Semaphore` in .NET Framework is using a different helper method to throw the exception and ends up throwing `IOException` instead.

In the other types I believe `DirectoryNotFoundException` would be thrown for names like "foo\\foo" or "global\\foo" (Global is case-sensitive). Other invalid names may return a different error code and result in a different exception. On Unix for named mutexes the error checking currently does not differentiate between cases in Windows that return `ERROR_PATH_NOT_FOUND`, `ERROR_INVALID_NAME`, AND `ERROR_INVALID_HANDLE`, as it was not attempted to replicate exactly the error checking behavior on Windows, and there any invalid name would result in `ERROR_INVALID_NAME` currently, leading to a `WaitHandleCannotBeOpenedException` for `Open`/`TryOpen` and `IOException` for `Create` and constructors. It seems confusing, with different exceptions being thrown for different types of invalid names and same exception being thrown for invalid name or some other reason.

Reply:
Ok great, thank you. I wrote a unit test that throws `DirectoryNotFoundException` when using `global\foo` as the name.